### PR TITLE
Code refactoring

### PR DESCRIPTION
**Pull request by shuding**
There are many code smells and this PR tries to clean up some of them. Also I expect the loader changes will speed up compilation a bit when the project is large.

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] [e2e](https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md#writing-tests-for-nextjs) tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have a helpful link attached, see [`contributing.md`](https://github.com/vercel/next.js/blob/canary/contributing.md)

## Documentation / Examples

- [ ] Make sure the linting passes by running `pnpm build && pnpm lint`
- [ ] The "examples guidelines" are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing/examples/adding-examples.md)

### CODE REVIEW

File: packages/next/src/build/index.ts [hunk 1]
```diff
@@ -2196,11 +2196,7 @@ export default async function build(
             exportPageWorker: sharedPool
               ? staticWorkers.exportPage.bind(staticWorkers)
               : undefined,
-            endWorker: sharedPool
-              ? async () => {
-                  await staticWorkers.end()
-                }
-              : undefined,
```
Comment (shuding) [inline packages/next/src/build/index.ts:2203]: Note that `staticWorkers.end()` doesn't take any argument so this change should be safe.

File: packages/next/src/build/webpack-config.ts [hunk 2]
```diff
@@ -1121,6 +1109,11 @@ export default async function getBaseWebpackConfig(
   const optOutBundlingPackages = EXTERNAL_PACKAGES.concat(
     ...(config.experimental.serverComponentsExternalPackages || [])
   )
+  const optOutBundlingPackageRegex = new RegExp(
+    `[/\\\\]node_modules[/\\\\](${optOutBundlingPackages
+      .map((p) => p.replace(/\//g, '[/\\\\]'))
+      .join('|')})[/\\\\]`
+  )
```
Comment (shuding) [inline packages/next/src/build/webpack-config.ts:1116]: Using RegExp as loader conditions (that runs over **all** imports) should be faster than function calls with multiple `includes` when scaled up.

File: packages/next/src/build/webpack-config.ts [hunk 3]
```diff
@@ -1752,24 +1745,18 @@ export default async function getBaseWebpackConfig(
...
-                issuerLayer: WEBPACK_LAYERS.server,
-                use: {
-                  loader: 'next-flight-loader',
+                issuerLayer: {
+                  or: [
+                    WEBPACK_LAYERS.server,
+                    WEBPACK_LAYERS.client,
+                    WEBPACK_LAYERS.appClient,
+                  ],
                 },
```
Comment (shuding) [inline packages/next/src/build/webpack-config.ts:1759]: This loader can be merged into the other one.

File: packages/next/src/build/webpack-config.ts [hunk 4]
```diff
@@ -1784,61 +1771,15 @@ export default async function getBaseWebpackConfig(
...
-                      // when react is acting as dependency of compiled/react-dom.
-                      alias: {
-                        react: 'next/dist/compiled/react/react.shared-subset',
-                        // Use server rendering stub for RSC
-                        // x-ref: https://github.com/facebook/react/pull/25436
-                        'react-dom$':
-                          'next/dist/compiled/react-dom/server-rendering-stub',
-                      },
-                    },
-                  },
```
Comment (shuding) [inline packages/next/src/build/webpack-config.ts:1816]: This loader seems redundant.

File: packages/next/src/build/webpack-config.ts [hunk 5]
```diff
@@ -1784,61 +1771,15 @@ export default async function getBaseWebpackConfig(
...
-                      },
-                    },
+                issuerLayer: WEBPACK_LAYERS.client,
+                test: codeCondition.test,
+                resolve: {
+                  alias: {
+                    react: 'next/dist/compiled/react',
+                    'react-dom$':
+                      'next/dist/compiled/react-dom/server-rendering-stub',
                   },
```
Comment (shuding) [inline packages/next/src/build/webpack-config.ts:1781]: Not sure about this loader condition, maybe it's for `appClient`?

File: packages/next/src/server/app-render.tsx [hunk 6]
```diff
@@ -315,9 +316,6 @@ function useFlightResponse(
   if (flightResponseRef.current !== null) {
     return flightResponseRef.current
   }
-  const {
-    createFromReadableStream,
-  } = require('next/dist/compiled/react-server-dom-webpack/client')
```
Comment (shuding) [inline packages/next/src/server/app-render.tsx:320]: No need to require it in every render pass.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: b5704c33f758ec4af6c77519a689cc7ccbfd851e

`pnpm testheadless test/integration/amphtml-fragment-style/test/index.test.js` 
- AMP Fragment Styles > adds styles from fragment in AMP mode correctly

<details>
<summary>Expand output</summary>

  ● AMP Fragment Styles › adds styles from fragment in AMP mode correctly

    command failed with code 1 signal null
    info  - Linting and checking validity of types...
    warn  - The Next.js plugin was not detected in your ESLint configuration. See https://nextjs.org/docs/basic-features/eslint#migrating-existing-config
    info  - Creating an optimized production build...
    info  - Using locally built binary of @next/swc
    info  - Compiled successfully
    info  - Collecting page data...
    info  - Generating static pages (0/3)
    info  - Generating static pages (3/3)

    > Build error occurred
    TypeError: Cannot read properties of undefined (reading '_worker')

      56 |     }
      57 |     end() {
    > 58 |         const worker = this._worker;
         |                             ^
      59 |         if (!worker) {
      60 |             throw new Error("Farm is ended, no more calls can be done to it");
      61 |         }

      at end (../packages/next/dist/lib/worker.js:58:29)
      at ../packages/next/dist/export/index.js:377:34
      at async Span.traceAsyncFn (../packages/next/dist/trace/trace.js:79:20)
      at ../async /work/packages/next/dist/build/index.js:1372:21
      at async Span.traceAsyncFn (../packages/next/dist/trace/trace.js:79:20)
      at ../async /work/packages/next/dist/build/index.js:1232:17
      at async Span.traceAsyncFn (../packages/next/dist/trace/trace.js:79:20)
      at async Object.build [as default] (../packages/next/dist/build/index.js:71:29)
      at ChildProcess.<anonymous> (lib/next-test-utils.js:258:11)


  ● Test suite failed to run

    TypeError: Cannot read properties of undefined (reading 'pid')

      463 | // Kill a launched app
      464 | export async function killApp(instance) {
    > 465 |   await killProcess(instance.pid)
          |                              ^
      466 | }
      467 |
      468 | export async function startApp(app) {

      at killApp (lib/next-test-utils.js:465:30)
      at integration/amphtml-fragment-style/test/index.test.js:24:25

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Pull request closed without merging**